src/m365/spfx/commands/project/project-azuredevops-pipeline-add.ts

```diff
@@ -0,0 +1,213 @@
+    }
+    catch (error: any) {
+      throw new CommandError(error);
+    }
+  }
+
+  private savePipeline(pipeline: azureDevOpsPipeline): void {
+    const azureDevOpsPath: string = path.join(this.projectRootPath as string, '.azuredevops');
+    fsUtil.ensureDirectory(azureDevOpsPath);
+
+    const pipelinesPath: string = path.join(azureDevOpsPath, 'pipelines');
+    fsUtil.ensureDirectory(pipelinesPath);
+
+    const pipelineFile: string = path.join(pipelinesPath, 'deploy-spfx-solution.yml');
+    fs.writeFileSync(path.resolve(pipelineFile), yaml.stringify(pipeline), 'utf-8');
+  }
+
+  private updatePipeline(solutionName: string, pipeline: azureDevOpsPipeline, options: GlobalOptions): void {
+    if (options.name) {
+      pipeline.name = options.name;
+    }
+    else {
+      delete pipeline.name;
+    }
+
+    if (options.branchName) {
+      pipeline.trigger.branches.include[0] = options.branchName;
+    }
+
+    const version = this.getProjectVersion();
+
+    if (!version) {
+      throw `Unable to determine the version of the current SharePoint Framework project`;
+    }
+
+    const match = version.match(/^(\d+)\.(\d+)\.(\d+)$/);
+    const minorVersionString = match ? match[2] : null;
+    const minorVersion = minorVersionString ? Number(minorVersionString) : null;
+
+    if (minorVersion === null || isNaN(minorVersion)) {
+      throw `Unable to determine the minor version of the current SharePoint Framework project`;
+    }
+
+    if (minorVersion < 18) {
+      this.getNodeAction(pipeline).inputs!.versionSpec = '16.x';
+    }
+
+    if (options.loginMethod === 'user') {
+      const script = this.getScriptAction(pipeline);
+      script.script = script.script!.replace(`m365 login --authType certificate --certificateBase64Encoded '$(CertificateBase64Encoded)' --password '$(CertificatePassword)' --appId '$(EntraIDAppId)' --tenant '$(TenantId)'`, `m365 login --authType password --userName '$(UserName)' --password '$(Password)'`);
+      pipeline.variables = pipeline.variables.filter(v => v.name !== 'CertificateBase64Encoded');
+      pipeline.variables = pipeline.variables.filter(v => v.name !== 'CertificateSecureFileId');
+      pipeline.variables = pipeline.variables.filter(v => v.name !== 'CertificatePassword');
+      pipeline.variables = pipeline.variables.filter(v => v.name !== 'EntraIDAppId');
+      pipeline.variables = pipeline.variables.filter(v => v.name !== 'TenantId');
+    }
+    else {
+      pipeline.variables = pipeline.variables.filter(v => v.name !== 'UserName');
+      pipeline.variables = pipeline.variables.filter(v => v.name !== 'Password');
+    }
+
+    if (options.scope === 'sitecollection') {
+      const script = this.getScriptAction(pipeline);
+      script.script = script.script!.replace(`m365 spo app deploy --name '$(PackageName)' --appCatalogScope 'tenant'`, `m365 spo app deploy --name '$(PackageName)' --appCatalogScope sitecollection --appCatalogUrl '$(SiteAppCatalogUrl)'`);
```

Comment:
If we change anything about the script in the pipeline definition, this will fail. Isn't there a better way to do this, eg. define a token in the script in the pipeline definition which we replace here?